### ISSUE TITLE
PSX HW: Severe pixelation across rendered graphics

### ISSUE THREAD

**Issue by davidhedlund** [libretro/beetle-psx-libretro#900]
I did a quick search but couldn't locate an existing report using specific graphics terminology.  If a similar report exists, please feel free to link it here and close this issue.

# Beetle PSX HW
I'm experiencing visual artifacts across the entire display. The sky, in particular, exhibits a noticeable pixelation effect.

![Screenshot at 2024-06-12 23-41-53](https://github.com/libretro/beetle-psx-libretro/assets/6416073/a279053b-4763-4ef1-aeef-ca1a8d9e0e4c)

# Beetle PSX
No pixelation issues like with Beetle PSX HW.

![Screenshot at 2024-06-12 23-44-59](https://github.com/libretro/beetle-psx-libretro/assets/6416073/00a79278-eb12-4476-b750-9eb812c98785)

**Comment by hizzlekizzle** [comment]
Yeah, the OpenGL renderer has a lot of bugs, unfortunately. The Vulkan renderer is much closer to software rendering in reliability.

**Comment by davidhedlund** [comment]
"Beetle PSX is built without hardware rendering support and lacks enhancements such as texture filtering, PGXP perspective correct texturing, and more. To take advantage of hardware rendering and the various enhancements that come with it, use Beetle PSX HW instead." - https://docs.libretro.com/library/beetle_psx/ 

> Yeah, the OpenGL renderer has a lot of bugs, unfortunately. The Vulkan renderer is much closer to software rendering in reliability.

Thank you! I think that should be added to https://github.com/libretro/docs/blob/master/docs/library/beetle_psx.md -- What do you think about it?

**Comment by hizzlekizzle** [comment]
Sure, fine by me.

**Comment by davidhedlund** [comment]
> Sure, fine by me.

Thanks:
* https://github.com/libretro/docs/pull/950

**Comment by mudl0rd** [comment]
> Yeah, the OpenGL renderer has a lot of bugs, unfortunately. The Vulkan renderer is much closer to software rendering in reliability.

Wonder how much effort it would take to rewrite it.

**Comment by hizzlekizzle** [comment]
@mudl0rd I would think that would depend on whether you wanted to fix the OpenGL renderer, backport the Vulkan renderer to OGL or introduce an abstraction so they share the same backend (ala Duck/Swanstation).

**Comment by mudl0rd** [comment]
Well I saw one major thing missing was any semblance of RTT, which is at least needed for FF battleswirl, MGS, etc.

**Comment by mudl0rd** [comment]
Managed to reproduce the issue at least with one of my fave games: THPS2. I noticed a VRAM read command was missing from the GL renderer which the VK renderer has - prob related to RTT? glReadPixels in the meantime could be used although that can be slow as heck on older GPUs like AGP.

**Comment by davidhedlund** [comment]
> Managed to reproduce the issue at least with one of my fave games: THPS2. Although I'll try some others where its more noticable.
> 
> I noticed a VRAM read command was missing from the GL renderer which the VK renderer has - prob related to RTT? glReadPixels with PBOs in the meantime could be used there although that can be slow as heck on older GPUs like AGP. The renderer itself would need to be completely reworked to render to framebuffers instead of the current libretro FB.
> 
> ![Screenshot 2024-10-14 075723](https://private-user-images.githubusercontent.com/158626984/376069368-6a9e0816-2949-4c67-bcf5-79e1e76913d2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg4OTk2MzUsIm5iZiI6MTcyODg5OTMzNSwicGF0aCI6Ii8xNTg2MjY5ODQvMzc2MDY5MzY4LTZhOWUwODE2LTI5NDktNGM2Ny1iY2Y1LTc5ZTFlNzY5MTNkMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxNFQwOTQ4NTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zOWM2YmUzOGJmMjlhNzM5OGYwMzQ2N2Q3MTFmMzYwZWFkNjRjOWY2ODA3YzIzZjg4MTNhYzczN2FhOTAwMTMyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.l-X93kYRa0bqTRTLWBxSBi0V7lsSwQ6rsABPnietdeM)

Thank you so much for taking the time to look into this and for all your hard work! I truly appreciate your efforts. I completely understand that this might be quite a challenge for your team, and I have every confidence in your abilities to tackle it. If there's anything I can do to help along the way, like conducting some simple testing, please don't hesitate to reach out.

**Comment by LibretroAdmin** [comment]
Check if this issue is still ongoing. I have plausible reason to believe it no longer does. If it does, please reopen it

**Comment by davidhedlund** [comment]
> Check if this issue is still ongoing. I have plausible reason to believe it no longer does. If it does, please reopen it

* Main Menu → Online Updater → Update Installed Cores
  - Updates to version 0.9.44.1 a98c7e3
* This issue (note that I reported this as OpenGL issue) has still not been fixed

Quick Settings -> Core Options -> Video
* Internal GPU Resolution: 1x (Native)
* Internal Color Dept: 16 bpp (Native)

<img width="901" height="770" alt="Image" src="https://github.com/user-attachments/assets/d2a33bd8-3cbe-4538-a089-070456830c05" />

Please re-open or give me permission to re-open the issues that I've submitted.

**Comment by Starlightbotanist** [comment]
@davidhedlund Can you test again on the latest beetle HW core (there have been a few openGL changes the past week or so)? If it's still happening, I will try to setup a test environment this weekend and track this down.

In case it's not reproducible on my end, can you share some hardware details and info about your distro?

**Comment by davidhedlund** [comment]
> [@davidhedlund](https://github.com/davidhedlund) Can you test again on the latest beetle HW core (there have been a few openGL changes the past week or so)? If it's still happening, I will try to setup a test environment this weekend and track this down.

* Tested on RetroArch Nightly
* Main Menu → Online Updater → Update Installed Cores
  - Updates to version 0.9.44.1 ddd7884

Issue is still going on.

<img width="881" height="750" alt="Image" src="https://github.com/user-attachments/assets/378cee7b-0516-41c2-8f23-e621e4693a25" />

> In case it's not reproducible on my end, can you share some hardware details and info about your distro?

Please try to reproduce it yourself first, because I need to upgrade my distro.

**Comment by davidhedlund** [comment]
> Please try to reproduce it yourself first, because I need to upgrade my distro.

Here are all the settings that you need in a nutshell:

`Settings → Output → Video` when set to "gl" (default) or "glcore".

The most important settings to reproduce visible rendering issues are to quote the main post from the "Internal GPU Resolution: 4x. Internal Color Dept: 16 bpp (Native)" section:

> Quick Settings -> Core Options -> Video
> * Internal GPU Resolution: 4x
> * Internal Color Dept: 16 bpp (Native)
